File: CyberKnet.Foscam/Sockets/WebSocketServiceManager.cs
using CyberKnet.Foscam.Responses;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WebSocket4Net;

namespace CyberKnet.Foscam.Sockets
{
    internal class WebSocketServiceManager
    {
        private bool needServiceUpdate = false;
        private Camera Camera { get; set; }
        private int ManagementPort { get; set; }
        WebSocket Socket = null;

        private void log(string msg)
        {

        }

        public delegate void PortReceivedEventHandler(object sender, PortEventArgs e);
        public event EventHandler<PortEventArgs> PortReceived;

        public WebSocketServiceManager(Camera camera, int managementPort)
        {
            Camera = camera;
            ManagementPort = managementPort;
        }

        public void RequestService()
        {
            Socket = new WebSocket4Net.WebSocket($"ws://127.0.0.1:{ManagementPort}", "http", WebSocket4Net.WebSocketVersion.None);
            Socket.Opened += Socket_Opened;
            Socket.Closed += Socket_Closed;
            Socket.MessageReceived += Socket_MessageReceived;
            Socket.Error += Socket_Error;
        }

        protected virtual void OnRaisePortReceived(PortEventArgs e)
        {
            var handler = PortReceived;
            if (handler != null)
            {
                handler(this, e);
            }
        }

        private void Socket_Error(object sender, SuperSocket.ClientEngine.ErrorEventArgs e)
        {
            throw new NotImplementedException();
        }

        private void Socket_MessageReceived(object sender, MessageReceivedEventArgs e)
        {
            try
            {
                string json = getJsonFromMessage(e.Message);
                dynamic msg = JObject.Parse(json);
                if (json == null)
                    return;
                var msgid = msg.msgid * 1;
                //this.me.log("manager msgid : " + msgid);
                switch (msgid)
                {
                    case FoscamResponseType.RESPONSE_REQUEST_PORT:
                        var response = JsonConvert.DeserializeObject<PortResponse>(json);
                        OnRaisePortReceived(new PortEventArgs(response.Port));
                        //VersionCheck(json);
                        break;

                    default:
                        break;
                }

            }
            catch (Exception e) { }
        }

        private void Socket_Closed(object sender, EventArgs e)
        {
            log("Closed");
        }

        private void Socket_Opened(object sender, EventArgs e)
        {
            if (Socket.State == WebSocketState.Open)
            {
                Camera.RequestPort();
            }
        }

        private string getJsonFromMessage(object evt)
        {
            if (evt.data is ArrayBuffer)
            {
                var dataArray = new Int8Array(evt.data);
                var jsonStr = this.ab2str(dataArray);
                return jsonStr;
            }
            return null;
        }

        private string ab2str(byte[] buf)
        {
            return String.fromCharCode.apply(null, new Uint8Array(buf));
        }

        //private void downPlugins()
        //{
        //    var obj = document.getElementById("downPlugins");
        //    obj.innerHTML = '<a id="down" href="IPCWebComponents.exe">' + lg.get("IDS_LOGIN_KJ") + '</a>';
        //}

        //private void VersionCheck(PortResponse response)
        //{
        //    //check plugin version
        //    var vsIf = response.ServiceVersion;

        //    var v1 = vsIf.split(".");   //plugin version

        //    var v2 = pVersion.split(".");    //web version

        //    for (var m = 0; m < 4; m++)
        //    {
        //        if (v2[m] * 1 > v1[m] * 1)
        //        {
        //            needServiceUpdate = true;
        //            break;
        //        }
        //        else if (v2[m] * 1 == v1[m] * 1)
        //        {
        //            continue;
        //        }
        //        else
        //        {
        //            needServiceUpdate = false;
        //            break;
        //        }
        //    };
        //    if (needServiceUpdate)
        //    {
        //        if (IsFirefox())
        //        {
        //            var obj = document.getElementById("downPlugins");
        //            obj.innerHTML = '<a id="down" href="IPCWebComponents.exe">' + lg.get("IDS_PLUGIN_UPDATE_TIP") + '</a>';
        //        }
        //    }
        //}
    }
}

File: Javascript/websocket_service_manager.js
var wsServiceManager = new websocketServiceManager();
var needServiceUpdate = false;
function websocketServiceManager () {
	this.socket = null;
	this.log = function(msg) {
		try {
			console.log(msg);
		} catch (e) {}
	}
	this.rand = function () {
		var aDate = (new Date()).getTime()+"";
		return aDate.substring(aDate.length-9, aDate.length)*1;
	}
	
	this.requestService = function () {
		var str = window.location.href;
		var httpver = str.split("://")[0];
		var url = 'ws://127.0.0.1:' + SERVICE_MANAGER_PORT;  
		if(httpver == "https"){
			return;
		}
		this.socket = new WebSocket(url);
		this.socket.binaryType = "arraybuffer";
		this.socket.onopen = this.onOpen;
		this.socket.onmessage = this.onMessage;
		this.socket.onclose = this.onClose;
		this.socket.onerror = this.onError;
		this.socket.me = this;
		//this.log(this.socket);
	}

	this.onClose = function () {
		this.me.log("onClose");
	}

	this.onError = function () {
		this.me.log("onError");
		isPlugin = false;
		setTimeout(function() {
			DownServer();
		}, 1000);
		//this.me.downPlugins();
	}
	
	this.onOpen = function () {
		// this is this.socket!!!
		if (this.me.socket.readyState == 1) {
			this.me.requestPort();
		}
	}
	
	this.requestPort = function () {
		var data = new wsdata();
		data.version = 1;
		data.msgId = CMD_REQUEST_PORT;
		data.groupId = this.rand();
		data.sequence = 0;
		data.datalen = 0;
		var transData = JSON.stringify(data);
		this.socket.send(transData);
	}
	
	
	this.getJsonFromMessage = function (event) {
		if (event.data instanceof ArrayBuffer) {
			var dataArray = new Int8Array(event.data);
			var jsonStr = this.ab2str(dataArray);
			var jsonObject = JSON.parse(jsonStr);
			return jsonObject;
		}	
		return null;
	}
	this.ab2str = function (buf) {
		return String.fromCharCode.apply(null, new Uint8Array(buf));
	}
	this.onMessage = function (event) {
		try {
			var json = this.me.getJsonFromMessage(event);
			if (json == null)
				return;
			var msgid = json.msgid * 1;
			//this.me.log("manager msgid : " + msgid);
			switch (msgid) {
			case RESPONSE_REQUEST_PORT:{ //50000
				g_DstPort = json.dstPort;
				wsCore.connect(g_DstPort);
				this.me.VersionCheck(json);
				break;
			}
			default:
				break;
			}
			
		} catch (e) {}
	}
	this.downPlugins = function(){
		var obj = document.getElementById("downPlugins");
   		obj.innerHTML = '<a id="down" href="IPCWebComponents.exe">' + lg.get("IDS_LOGIN_KJ") + '</a>';
	}
	this.VersionCheck = function(json){
		//check plugin version
	    vsIf = json.seviceVer;
	    var v1 = vsIf.split(".");   //plugin version
	    var v2 = pVersion.split(".");    //web version
	    
	    for(var m = 0; m < 4; m++){
	        if(v2[m] * 1 > v1[m] * 1){
	            needServiceUpdate = true;
	            break;
	        }else if(v2[m] * 1 == v1[m] * 1){
	            continue;
	        }else{
	            needServiceUpdate = false;
	            break;
	        }
	    };
		if (needServiceUpdate) {
			if(IsFirefox()){
        		var obj = document.getElementById("downPlugins");
        		obj.innerHTML = '<a id="down" href="IPCWebComponents.exe">' + lg.get("IDS_PLUGIN_UPDATE_TIP") + '</a>';
			}
		}
	}
}